### DISCUSSION TITLE
Migrate external library content to internal library

### DISCUSSION THREAD

**Discussion by Sithuk** [immich-app/immich#12562] [General]
Is there a way to migrate an external library content to Immich's internal library so that all the assets (photos/videos) do not need to be re-scanned?

**Comment by blackangelpr** [comment]
i have the same question moved away from hosting on nextcloud locally to immich for the photos so its easier for the families wondering how to move their user photos to immich

**Comment by Sithuk** [comment]
I have added my photos into Immich as an external library. I don't want to remove the external library and then have to re-import them to the internal library. I am hoping there is a way to migrate the photos across so I don't have to re-scan everything again.

**Comment by rhyst** [comment]
I also would like this, even if its a guide for a manual process. I started using Immich a while ago with an external library because I liked to manage tagging. Now that tags are manageable in Immich I would prefer to move all my photos into the normal Immich library so they share the same folder structure. 

I could just re-upload with the CLI (I don't mind too much about waiting for everything to re-process) but the effort put into face recognition and tagging the faces correctly would be lost I imagine?

**Comment by rhyst** [comment]
I made a POC: https://github.com/immich-app/immich/pull/12724. Potentially using the template migration code is too much of a hack. Will see if it gets feedback.

**Comment by xtbqimuxmvtx** [comment]
Is this feature still planned? Is the external library feature still under a feature freeze?

I havent looked very far but is there any tool or plan to be able to write XMP sidecar data into the media files where compatible? I like external media so I know there's a raw untouched copy of the source files but then I do irregular maintainance to the external directiories and updating date taken would be super useful

**Reply by bo0tzz** [reply]
The intent is to eventually unify everything so there is no longer the concept of an "external" library.

> any tool or plan to be able to write XMP sidecar data into the media files

You can do this with exiftool.

**Reply by fabb** [reply]
> The intent is to eventually unify everything so there is no longer the concept of an "external" library.

That's very interesting. Is this still planned?

**Reply by bo0tzz** [reply]
Yes

**Comment by topongo** [comment]
hi, i had a similar issue for a long time and this past days i decided to give it a try. here's a rough guide on how to do it

1. i was (probably recklessly) confident in this amazing piece of software that's immich: i didn't backed up  my stuff, also because i'm running out of disk space. but please do it if you're storing critical data.
2. first i imported all my external assets into my library using [immich-go](https://github.com/simulot/immich-go)
    - got an api key
    - navigated to the external library path
    - imported all the assets: `immich-go upload from-folder --server http://localhost:2283 --api-key <redacted> .`
    - that took a while.
    - now you have a bunch of duplicated assets: that's normal. immich (and immich-go, consequentially) usually detects duplicates and skips unneeded uploads, but since your original assets are in an external library this feature is intentionally disabled.
4. next i generated a rust openapi client (you can use any language really):
    - to do it i used [`openapi-generator-cli`](https://github.com/OpenAPITools/openapi-generator-cli) with the [immich's openapi specification](https://github.com/immich-app/immich/blob/main/open-api/immich-openapi-specs.json). you can install `openapi-generator-cli` on your system or simply use docker:
    ```sh
    docker run --rm \
        -v $PWD:/local openapitools/openapi-generator-cli generate \
        -i /local/immich-openapi-specs.json \
        -g <PROGRAMMING_LANGUAGE_HERE> \
        -o /local/immich_api
    ```
    - this will create a folder `immich_api` that you can import as a rust crate (or python module, or whatever) and write your application with.
5. next i wrote a ~~simple~~ script that:
    - scans all my assets
    - pairs the external assets and the internal (freshly uploaded) ones that, in theory, are identical.
        - i tried using the `checksum` field, but it isn't generated by simply hashing the file data, so it's useless for our purposes.
        - i ended up using the `original_file_name` field: since all the external assets had a timestamp in the name, it actually uniquely identifies all my external assets, but your experience may vary.
        - i made sure not to have double duplicates (two files with the same name in the same library (external or internal)).
        - for distinguish the assets from internal and external i avoided using the `library_id` field, since apparently it's deprecated. i instead used the `original_path`, making sure that that starts with `/mnt/photos` (the path inside the container in which is mounted the external library).
    - then it scans all the albums and it makes sure that if `Album 1` contained assets from the external library their "internal library" counterparts are added to `Album 1`.
    for example: `Album 1` contains `/library/asset1.jpg` and `/mnt/photos/asset2.jpg`. the program finds the pair `("/mnt/photos/asset2.jpg", "/library/asset2.jpg")` (that are identical), and adds `/library/asset2.jpg` to `Album 1`
    - it does the same for all favorites assets
6. i don't know if this is necessary but you may want to run the `Face Detection` and the `Face Recognition` jobs before unplugging the external library for making sure the new assets are associated with the persons you have in your db.
7. in theory it should be now safe to unplug the external library.

**Comment by pier-gui** [comment]
I wrote a Python script to move my external assets to internal the way I wanted. It uses immich API endpoints. It will preserve metadata (edits to timestamps, tags, favourited status, albums), but thumbnails and machine learning do need to be re-generated for each asset.

https://github.com/pier-gui/Immich-external-library-to-internal
